SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/Components/SpatialPingComponent.h"
+
+#include "GameFramework/PlayerController.h"
+#include "Net/UnrealNetwork.h"
+#include "TimerManager.h"
+#include "Engine/World.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialPingComponent);
+
+USpatialPingComponent::USpatialPingComponent(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+{
+	PrimaryComponentTick.bCanEverTick = false;
+	PrimaryComponentTick.bStartWithTickEnabled = false;
+
+	bReplicates = true;
+}
+
+void USpatialPingComponent::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME_CONDITION(USpatialPingComponent, ReplicatedPingID, COND_OwnerOnly);
+}
+
+void USpatialPingComponent::BeginPlay()
+{
+	Super::BeginPlay();
+	// Attempt to cast the owner of this component to the PlayerController class.
+	// Component will do nothing if cast fails.
+	OwningController = Cast<APlayerController>(GetOwner());
+	if (OwningController == nullptr)
+	{
+		UE_LOG(LogSpatialPingComponent, Warning, TEXT("SpatialPingComponent did not find a valid owning PlayerController and will not function correctly. Ensure this component is only attached to a PlayerController."));
+	}
+
+	if (bStartWithPingEnabled)
+	{
+		SetPingEnabled(true);
+	}
+}
+
+void USpatialPingComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
+{
+	// Clear up the timer before the component is disabled or destroyed.
+	SetPingEnabled(false);
+	Super::EndPlay(EndPlayReason);
+}
+
+bool USpatialPingComponent::GetIsPingEnabled() const
+{
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	// Only execute if this component is attached to a player controller.
+	if (OwningController != nullptr)
+	{
+		// Only execute on owning local client.
+		if (OwningController->GetLocalRole() == ENetRole::ROLE_AutonomousProxy)
+		{
+			if (bSetEnabled && !bIsPingEnabled)
+			{
+				EnablePing();
+			}
+			else if (!bSetEnabled && bIsPingEnabled)
+			{
+				DisablePing();
+			}
+		}
+	}
+}
+
+float USpatialPingComponent::GetPing() const
+{
+	return RoundTripPing;
+}
+
+void USpatialPingComponent::EnablePing()
+{
+	if (UWorld* World = GetWorld())
+	{
+		LastSentPingID = 0;
+		TimeoutCount = 0;
+		// Send a new ping, which will trigger a self-perpetuating sequence via timers.
+		SendNewPing();
+		// Set looping timer to 'tick' this component, it doesn't send any pings but matches the MinPingInterval for passing updates to the owning controller.
+		World->GetTimerManager().SetTimer(PingTickHandle, this, &USpatialPingComponent::TickPingComponent, MinPingInterval, true);
```

Comment:
Setting the timer on the World's timer manager means time dilation will affect it if enabled.
We may not need to change this, but it may be worth printing a warning if the time dilation on the world isn't 1.

Reply:
Added a check and warning log message when enabling ping if time dilation is not 1